cardano-testnet/src/Testnet/Components/TestWatchdog.hs

```diff
@@ -0,0 +1,137 @@
+  , runWithDefaultWatchdog_
+  , runWithDefaultWatchdog
+  , Watchdog
+  , kickWatchdog
+  , poisonWatchdog
+  ) where
+
+import           Control.Concurrent (myThreadId, threadDelay, throwTo)
+import           Control.Concurrent.STM (atomically)
+import           Control.Concurrent.STM.TChan (TChan, newTChanIO, tryReadTChan, writeTChan)
+import           Control.Exception.Safe (Exception)
+import           Control.Monad.IO.Class
+import           Data.Time (NominalDiffTime, UTCTime, diffUTCTime, getCurrentTime,
+                   nominalDiffTimeToSeconds)
+import           GHC.Conc (ThreadId)
+import           GHC.Stack
+
+import qualified Hedgehog.Extras as H
+
+-- | Configuration for the watchdog.
+newtype WatchdogConfig = WatchdogConfig
+  { watchdogTimeout :: Int -- ^ Timeout in µs after which watchdog will kill the test case
+  }
+
+-- | Default watchdog config with 10 minutes timeout.
+defaultWatchdogConfig :: WatchdogConfig
+defaultWatchdogConfig = WatchdogConfig
+  { watchdogTimeout = 600_000_000
+  }
+
+-- | A watchdog
+data Watchdog = Watchdog
+  { watchdogConfig :: !WatchdogConfig
+  , watchedThreadId :: !ThreadId -- ^ monitored thread id
+  , startTime :: !UTCTime -- ^ watchdog creation time
+  , kickChan :: TChan WatchdogCommand -- ^ a queue of watchdog commands
+  }
+
+-- | Create a new watchdog
+makeWatchdog :: MonadIO m
+             => WatchdogConfig
+             -> ThreadId -- ^ thread id which will get killed after timeouts expire
+             -> m Watchdog
+makeWatchdog config watchedThreadId' = liftIO $ do
+  watchdog <- Watchdog config watchedThreadId' <$> getCurrentTime <*> newTChanIO
+  kickWatchdog watchdog
+  pure watchdog
+
+-- | Run watchdog in a loop
+runWatchdog :: MonadIO m
+            => Watchdog
+            -> m ()
+runWatchdog w@Watchdog{watchedThreadId, startTime, kickChan} = liftIO $ do
+  atomically (tryReadTChan kickChan) >>= \case
+    Just PoisonPill ->
+      -- deactivate watchdog
+      pure ()
+    Just (Kick timeout) -> do
+      -- got a kick, wait for another period
+      threadDelay timeout
+      runWatchdog w
+    Nothing -> do
+      -- we are out of scheduled timeouts, kill the monitored thread
+      currentTime <- getCurrentTime
+      throwTo watchedThreadId . WatchdogException $ diffUTCTime currentTime startTime
+
+-- | Watchdog command
+data WatchdogCommand
+  = Kick !Int -- ^ Add another delay
+  | PoisonPill -- ^ Stop the watchdog
+
+-- | Enqueue a kick for the watchdog. It will extend the timeout by another one defined in the watchdog
+-- configuration.
+kickWatchdog :: MonadIO m => Watchdog -> m ()
+kickWatchdog Watchdog{watchdogConfig=WatchdogConfig{watchdogTimeout}, kickChan} = liftIO $
+  atomically $ writeTChan kickChan (Kick watchdogTimeout)
+
+-- | Enqueue a poison pill for the watchdog. It will stop the watchdog after all timeouts.
+poisonWatchdog :: MonadIO m => Watchdog -> m ()
+poisonWatchdog Watchdog{kickChan} = liftIO $
+  atomically $ writeTChan kickChan PoisonPill
```

Comment:
Since this function is unused, I would KISS and remove it altogether. The same applies to `PoisonPill`. My rationale is that otherwise we're introducing unnecessary maintenance duties.

But maybe this is subjective, do like you feel is best; this is just a suggestion :slightly_smiling_face:

Reply:
I would like to leave it to provide more usable API for watchdog: for a user to be able to control it in a minimal level. In general I'm considering upstreaming it to hedgehog-extras since it's not very specific to our tests.

Reply:
@carbolymer> makes sense, so resolving :+1: